### PR TITLE
Changes in FreeRTOS+TCP demo application files to add compatibility for running EndPoint changes in the IPv4 stack.

### PR DESCRIPTION
**Pull request by kar-rahul-aws**
Changes in FreeRTOS+TCP demo application files to add compatibility for running EndPoint changes in the TCP stack.

Description
-----------
1. FreeRTOS_GetAddressConfiguration function is updated to **FreeRTOS_GetEndPointConfiguration** for supporting EndPoint changes
2. **FREERTOS_PLUS_TCP_VERSION** macro support is added to verify, which functionality to be used , with or without the EndPoint changes.
3. **vApplicationIPNetworkEventHook** function is added to UDP demo, to determine which demo project to run .
4.  Added **IPType** as parameter to **FreeRTOS_GetUDPPayloadBuffer** function, to be in sync with EndPoint changes 

Test Steps
-----------
Normal Demo run steps

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by shubnil — DISMISSED** [review]

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Posix/main_networking.c
Comment (moninom1) [inline FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Posix/main_networking.c:159]: Do we need TCP version flags here as well?

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Posix/main_networking.c
Comment (moninom1) [inline FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Posix/main_networking.c:172]: Please remove the commented part.

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Minimal_Windows_Simulator/main.c
Comment (moninom1) [inline FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Minimal_Windows_Simulator/main.c:169]: Might need to run uncrustify, indentation seems incorrect.

**Comment by moninom1** [comment]
Update the commit message.

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Posix/main_networking.c
Comment (kar-rahul-aws) [reply FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Posix/main_networking.c:159]: Currently, we are adding the TCP flag only for the EndPoint change, the **FreeRTOS IPInit and IPStart** , can be defined by the **compatible with single** macro only, I think

Comment (shubnil) [inline FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Posix/main_networking.c:159]: Ideally, we do not want multiple macros. This part of the code also has endpoint changes like FreeRTOS_FillEndPoint, so using a single macro makes more sense. We can use FREERTOS_PLUS_TCP_VERSION in all the places.

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Posix/main_networking.c
Comment (shubnil) [inline FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Posix/main_networking.c:121]: Ideally, we do not want multiple macros. This part of the code also has endpoint changes like "static NetworkEndPoint_t xEndPoints[4];", so using a single macro makes more sense. We can use FREERTOS_PLUS_TCP_VERSION in all the places.

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Qemu_mps2/main_networking.c
Comment (shubnil) [inline FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Qemu_mps2/main_networking.c:126]: same as above.

**Review by shubnil — CHANGES_REQUESTED** [review]

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Posix/main_networking.c
Comment (shubnil) [reply FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Posix/main_networking.c:121]: I basically want to understand the use case where adding 2 extra macros ipconfigMULTI_INTERFACE and ipconfigCOMPATIBLE_WITH_SINGLE will make the demo more configurable or more useful, otherwise adding extra macros is only going to add up to more maintainability related issues. From the new version of FREERTOS_PLUS_TCP_VERSION >= 10, it is always multiple interface by default, so it makes more sense to have a single macro.

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Qemu_mps2/main_networking.c
Comment (kar-rahul-aws) [reply FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Qemu_mps2/main_networking.c]: Updated

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Qemu_mps2/main_networking.c
Comment (shubnil) [inline FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Qemu_mps2/main_networking.c:160]: Is this memcpy common to both old and new

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Qemu_mps2/main_networking.c
Comment (shubnil) [inline FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Qemu_mps2/main_networking.c:148]: Why are we removing the prints here.

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Qemu_mps2/main_networking.c
Comment (kar-rahul-aws) [reply FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Qemu_mps2/main_networking.c:160]: Yes , this memcpy is common to both old and new, since with EndPoint changes, we call **FreeRTOS_FillEndPoint**which requires a initial MAC address from the ipLOCAL_MAC_ADDRESS.

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Qemu_mps2/main_networking.c
Comment (kar-rahul-aws) [reply FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Qemu_mps2/main_networking.c:148]: Debug prints are not removed, they are added before the macro FREERTOS_PLUS_TCP_VERSION  is called in the Initialisation flow.

**Review by shubnil — APPROVED** [review]

**Comment by shubnil** [comment]
Please run the change on main and dev/ipv6_integration branches before merge.

**Comment by shubnil** [comment]
The change looks good to me. I can approve once we have the demo run results.

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Qemu_mps2/main_networking.c
Comment (moninom1) [inline FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Qemu_mps2/main_networking.c:172]: Please remove "// pdFALSE; // pdTRUE;"

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_UDP_Mode_CLI_Windows_Simulator/DemoTasks/CLI-commands.c
Comment (moninom1) [inline FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_UDP_Mode_CLI_Windows_Simulator/DemoTasks/CLI-commands.c:536]: ulIPAddress - no ulIPAddress in this function.

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_UDP_Mode_CLI_Windows_Simulator/DemoTasks/CLI-commands.c
Comment (moninom1) [inline FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_UDP_Mode_CLI_Windows_Simulator/DemoTasks/CLI-commands.c:547]: ulIPAddress - no ulIPaddress, need to replace it with ulAddress

Comment (moninom1) [reply FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_UDP_Mode_CLI_Windows_Simulator/DemoTasks/CLI-commands.c:547]: Need to be done in all places in this file

File: FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Qemu_mps2/main_networking.c
Comment (kar-rahul-aws) [reply FreeRTOS-Plus/Demo/FreeRTOS_Plus_TCP_Echo_Qemu_mps2/main_networking.c]: done

File: lexicon.txt
Comment (aggarg) [inline lexicon.txt:2615]: Is that not a typo?

**Review by aggarg — APPROVED** [review]

**Review by aggarg — APPROVED** [review]

**Comment by kar-rahul-aws** [comment]
The demos(WinSim, UDP, QEMU and Posix) are working with these changes

**Review by shubnil — DISMISSED** [review]

**Review by moninom1 — DISMISSED** [review]

**Review by moninom1 — APPROVED** [review]

**Pull request merged**
